sim-lib/src/lib.rs

```diff
@@ -604,6 +759,68 @@ async fn produce_events(
     shutdown.trigger();
 }
 
+async fn produce_random_events<N: NetworkGenerator, A: PaymentGenerator + Display>(
+    source: PublicKey,
+    network_generator: Arc<Mutex<N>>,
+    node_generator: A,
+    sender: Sender<SimulationEvent>,
+    shutdown: Trigger,
+    listener: Listener,
+) {
+    log::info!("Started random activity producer for {source}: {node_generator}.");
+
+    loop {
+        let wait = node_generator.next_payment_wait();
+        log::debug!("Next payment for {source} in {:?} seconds.", wait);
+
+        select! {
+            biased;
+            _ = listener.clone() => {
+                log::debug!("Random activity generator for {source} received signal to shut down.");
+                break;
+            },
+
+            // Wait until our time to next payment has elapsed then execute a random amount payment to a random
+            // destination.
+            _ = time::sleep(wait) => {
+                let destination = network_generator.lock().await.sample_node_by_capacity(source);
+
+                // Only proceed with a payment if the amount is non-zero, otherwise skip this round. If we can't get
+                // a payment amount something has gone wrong (because we should have validated that we can always
+                // generate amounts), so we exit.
+                let amount = match node_generator.payment_amount(destination.1) {
+                    Ok(amt) => {
+                        if amt == 0 {
+                            log::debug!("Skipping zero amount payment for {source} -> {}.", destination.0);
+                            continue;
+                        }
+
+                        amt
+                    },
+                    Err(e) => {
+                        log::error!("Could not get amount for {source} -> {}: {e}.", destination.0);
+                        break;
+                    },
```

Comment:
I'm OK with having this here, but this can actually never happen, right? We've sanitized the inputs to the simulation with `validate_capacity`

Reply:
Yeah, shouldn't happen.